waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -612,14 +612,23 @@ proc queryFromWithPaging*(w: WakuStore, query: HistoryQuery, peer: RemotePeerInf
   return ok(messageList)
 
 proc queryLoop(w: WakuStore, query: HistoryQuery, candidateList: seq[RemotePeerInfo]): Future[MessagesResult]  {.async, gcsafe.} = 
-  ## loops through the candidateList in order and sends the query to each until one of the query gets resolved successfully
-  ## returns the retrieved messages, or error if all the requests fail
-  for peer in candidateList.items: 
-    let successResult = await w.queryFromWithPaging(query, peer)
-    if successResult.isOk: return ok(successResult.value)
-
-  debug "failed to resolve the query"
-  return err("failed to resolve the query")
+  ## loops through the candidateList in order and sends the query to each
+  ## once all responses have been received, the retrieved messages are consolidated into one deduplicated list
+  ## if no messages have been retrieved, the returned future will resolve into a MessagesResult result holding an empty seq.
+  var futureList: seq[Future[MessagesResult]]
+  for peer in candidateList.items:
+    futureList.add(w.queryFromWithPaging(query, peer))
+  futureList = await allFinished(futureList) # TODO: is there a proc that returns a Future[seq[T]?
+  var messagesList: seq[WakuMessage]
+  # all futures have already been awaited, so the following loop runs without actually waiting
+  for fut in futureList:
+    let successResult = await fut
+    if successResult.isOk: messagesList.insert(successResult.value)
```

Comment:
I think it may be possible to map `futureList` to `messagesList` using the [`mapIt()`](https://nim-lang.org/docs/sequtils.html#mapIt.t%2Ctyped%2Cuntyped) template and `concat`ting the resulting sequence-of-sequences. Compressing too much logic into a single line is probably not a good idea, so feel free to ignore this suggestion, or just applying it to some steps. :)

Reply:
Thank you. I was looking for this, but then decided to take syntactically simple approach due to lack of Nim knowledge.
I will look into it an apply the change you suggested.